MCprep_addon/world_tools.py

```diff
@@ -110,51 +116,127 @@ def texture_type(self):
+	"mineways-c" : WorldExporter.ClassicMW,
+	"jmc2obj-c"  : WorldExporter.ClassicJmc
+}
+
+def get_exporter(context: Context) -> Optional[WorldExporter]:
+	"""
+	Return the exporter on the active object if it has 
+	an exporter attribute.
 
+	For maximum backwards compatibility, it'll convert the 
+	explicit options we have in MCprep for world exporters to 
+	WorldExporter enum objects, if the object does not have either
+	the CommonMCOBJ exporter attribute, or if it does not have the 
+	MCPREP_OBJ_EXPORTER attribute added in MCprep 3.6. This backwards
+	compatibility will be removed by default in MCprep 4.0
 
-def detect_world_exporter(filepath: Path) -> None:
+	Returns:
+		- WorldExporter if the world exporter can be detected
+		- None otherwise
+	"""
+	obj = context.active_object
+	if not obj:
+		return None
+
+	if "COMMONMCOBJ_HEADER" in obj:
+		if obj["exporter"] in EXPORTER_MAPPING:
+			return EXPORTER_MAPPING[obj["exporter"]]
+		else:
+			return WorldExporter.Unknown
+	elif "MCPREP_OBJ_HEADER" in obj:
+		if "MCPREP_OBJ_EXPORTER" in obj:
+			return EXPORTER_MAPPING[obj["MCPREP_OBJ_EXPORTER"]]
+	
+	# This section will be placed behind a legacy 
+	# option in MCprep 4.0, once CommonMCOBJ becomes
+	# more adopted in exporters
+	prefs = util.get_user_preferences(context)
+	if prefs.MCprep_exporter_type == "Mineways":
+		return WorldExporter.ClassicMW
+	elif prefs.MCprep_exporter_type == "jmc2obj":
+		return WorldExporter.ClassicJmc
+	return None
+
+def detect_world_exporter(filepath: Path) -> Union[CommonMCOBJ, ObjHeaderOptions]:
 	"""Detect whether Mineways or jmc2obj was used, based on prefix info.
 
 	Primary heruistic: if detect Mineways header, assert Mineways, else
 	assume jmc2obj. All Mineways exports for a long time have prefix info
 	set in the obj file as comments.
 	"""
+	obj_header = ObjHeaderOptions()
 	with open(filepath, 'r') as obj_fd:
 		try:
-			header = obj_fd.readline()
-			if 'mineways' in header.lower():
-				obj_header.set_mineways()
-				# form of: # Wavefront OBJ file made by Mineways version 5.10...
-				for line in obj_fd:
-					if line.startswith("# File type:"):
-						header = line.rstrip()  # Remove trailing newline
-
-				# The issue here is that Mineways has changed how the header is generated.
-				# As such, we're limited with only a couple of OBJs, some from
-				# 2020 and some from 2023, so we'll assume people are using
-				# an up to date version.
-				atlas = (
-					"# File type: Export all textures to three large images",
-					"# File type: Export full color texture patterns"
-				)
-				tiles = (
-					"# File type: Export tiles for textures to directory textures",
-					"# File type: Export individual textures to directory tex"
-				)
-				print('"{}"'.format(header))
-				if header in atlas:  # If a texture atlas is used
-					obj_header.set_atlas()
-				elif header in tiles:  # If the OBJ uses individual textures
-					obj_header.set_seperated()
-				return
+			cmc_header = parse_header(obj_fd)
```

Comment:
Also based on the implementation above, wouldn't the readlines() line within parse_header clear out the remainder of readlines below? Would be good to double check all tests still pass as you're developing here.

Reply:
Turns out this seems to be the source of the problem, indeed the parse_header above if commented out makes _more_ of the tests pass. Which also means we really do need to add also tests for the common obj part too, otherwise we're not really ensuring this continues to work.

Reply:
Through investigating this, I have also come to realize the existing commonmcobj parser actually always parses the _entire_ obj file, line by line, in search of the header. This could be a massive slowdown for large obj files.

FYI I'm looking into the best option to short circuit @StandingPadAnimations, most likely if we haven't seen the anchor `COMMON_MC_OBJ_START` within the first 100 lines of the header, then we're probably safe to say it's safe to exit early. Started this diff locally, will need to verify against actual Mineways and jmc2obj example commonmcOBJ exports when I get a chance:

```diff
    # Read in the header
-   for l in f:
-       tl = " ".join(l.rstrip().split())
-       if tl == "# COMMON_MC_OBJ_START":
+   lines_read = 0
+   for _l in f:
+       tl = " ".join(_l.rstrip().split())
+       lines_read += 1
+       if lines_read > 100 and tl and not tl.startswith("#"):
+           break  # no need to parse further than the true header area
        elif tl == "# COMMON_MC_OBJ_START":
            header.append(tl)
            found_header = True 
            continue
```

Reply:
CommonMCOBJ allows for headers to be anywhere in the file technically, either in the beginning or the end. That said, we could check for both 100 lines in the beginning and 100 lines at the end, and allow it to be configurable for users

Reply:
I [briefly looked](https://www.geeksforgeeks.org/python-reading-last-n-lines-of-a-file/) at efficiently reading the last N lines of a file without actually iterating over all of them. It's a little complicated. The other more practical consideration is that both jmc2obj and Mineways, as I've just verified, put commonobj towards the top. Currently in both exports I've done, the header starts ~line 55, and for jmc2obj it's at the very start. So, Seems like a buffer of 100 should be safe. I might time-test the difference later to see how much it would end up saving

Reply:
If we do implement a 100 line limit, then we should also allow the user to be able to configure it somewhere. That way, if an OBJ that is a CommonMCOBJ export doesn't put it in the first 100 lines, users wouldn't have to go down into the Python code and modify the limit.

I think a good way to test would be to have an OBJ file where the header is at the end (which should be trivial since OBJ is a text format) and benchmark the time difference in MCprep. Not only would it give us insights into whether or not it makes a practical difference, it would also test to see if any unexpected issues occur

Reply:
I'd be in favor of marking that a TODO if you're ok way it, to keep parsing simple and fast for now (otherwise it'll be seen as a regression).

Also spoiler I will not finish this tonight, I had to update and create a test world for v1.21 of MC and do exports for both jmc2obj and mineways. While that is working, it perhaps has revealed some issues with the mapping script which I am looking into still along with setting up tests to make sure parsing is working against real commonmcobj  world exports.

Reply:
Sounds good to me